predictive back gesture support for Android 13+

kubaflo opened the pull request:
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

<!--
!!!!!!! MAIN IS THE ONLY ACTIVE BRANCH. MAKE SURE THIS PR IS TARGETING MAIN. !!!!!!! 
-->

### Description of Change

This PR introduces a unified HandleBackNavigation method to centralize and standardize back navigation logic across platforms.

On Android 13 (API 33) and above, it integrates predictive back gesture callbacks with the MAUI lifecycle, ensuring that custom back navigation seamlessly participates in the system’s predictive back animations.

For more information on predictive back navigation, see the official Android documentation:
https://developer.android.com/guide/navigation/custom-back/predictive-back-gesture#update-custom

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes https://github.com/dotnet/maui/issues/32458
Fixes https://github.com/dotnet/maui/issues/32750

<!--
Are you targeting main? All PRs should target the main branch unless otherwise noted.
-->

kubaflo committed:
Add predictive back gesture support for Android 13+

Introduces a unified HandleBackNavigation method to centralize back navigation handling and integrates Android 13+ predictive back gesture callbacks with MAUI lifecycle events. Predictive back is registered and unregistered appropriately, ensuring custom back handling works with system back gesture animation.

mattleibow reviewed (commented):
## Thorough Review: Predictive Back Gesture Support for Android 13+

### Summary
This PR adds predictive back gesture support for Android 13+ (API 33+) by implementing the `OnBackInvokedCallback` API. It centralizes back navigation handling in a new `HandleBackNavigation()` method used by both the legacy `OnBackPressed()` and the new predictive back callback system.

The approach correctly integrates MAUI's lifecycle event system with Android's predictive back gestures while maintaining backward compatibility.

---

### Critical Issues

#### 🔴 **BLOCKING: Potential Double Registration of Callback**
**Location**: `MauiAppCompatActivity.cs`, lines 44-45

The code could register the same callback multiple times if `OnCreate()` is called repeatedly during the activity lifecycle (e.g., configuration changes, activity recreation).

**Current Code**:
```csharp
_predictiveBackCallback ??= new PredictiveBackCallback(this);
dispatcher?.RegisterOnBackInvokedCallback(0, _predictiveBackCallback);
```

**Issue**: `??=` prevents creating a new instance, but `RegisterOnBackInvokedCallback()` is called unconditionally, potentially registering the same callback multiple times.

**Required Fix**:
```csharp
if (_predictiveBackCallback is null)
{
    _predictiveBackCallback = new PredictiveBackCallback(this);
    dispatcher?.RegisterOnBackInvokedCallback(0, _predictiveBackCallback);
}
```

---

#### ⚠️ **HIGH: Missing Dispose Call**
**Location**: `MauiAppCompatActivity.cs`, `OnDestroy()` around line 53

`PredictiveBackCallback` inherits from `Java.Lang.Object` and should be explicitly disposed to prevent memory leaks.

**Current Code**:
```csharp
OnBackInvokedDispatcher?.UnregisterOnBackInvokedCallback(_predictiveBackCallback);
_predictiveBackCallback = null;
```

**Required Fix**:
```csharp
OnBackInvokedDispatcher?.UnregisterOnBackInvokedCallback(_predictiveBackCallback);
_predictiveBackCallback.Dispose();
_predictiveBackCallback = null;
```

---

### Design Discussion

#### Regarding @mattleibow's Question: Should `base.OnBackPressed()` be called from PredictiveBackCallback?

**Answer: Yes, this is the correct approach.**

**Reasoning**:
1. **MAUI Architecture**: MAUI uses `AndroidLifecycle.OnBackPressed` events to allow application code to intercept and customize back navigation behavior
2. **Platform Behavior**: `base.OnBackPressed()` in `AppCompatActivity` handles critical platform operations:
   - Fragment back stack management
   - Activity finishing
   - Integration with AndroidX navigation components
3. **Compatibility**: This design maintains existing MAUI navigation patterns while adding predictive back animation support
4. **Control Flow**:
   - User performs back gesture → `OnBackInvoked()` called
   - `HandleBackNavigation()` invokes MAUI lifecycle events
   - App code can set `preventBackPropagation = true` to handle custom back behavior
   - If not prevented → `base.OnBackPressed()` performs default platform navigation

This unified approach correctly bridges MAUI's lifecycle system with Android's predictive back gesture API while preserving backward compatibility with Android 12 and earlier.

---

### Minor Issues

#### ℹ️ **LOW: Hardcoded Priority Value**
**Location**: `MauiAppCompatActivity.cs`, line 45

The priority `0` is hardcoded. While correct (`PRIORITY_DEFAULT = 0`), adding a comment would improve clarity:

```csharp
// Priority 0 = PRIORITY_DEFAULT: callback invoked only when no higher-priority callback handles the event
dispatcher?.RegisterOnBackInvokedCallback(0, _predictiveBackCallback);
```

#### 📝 **NOTE: Pragma Warnings**
The pragma warnings (`CA1416`/`CA1422`) were removed from `OnBackPressed()` when the code moved to `HandleBackNavigation()`. These warnings suppress obsolete API alerts. While the warnings aren't critical and were part of older Android code patterns, they could be re-added to `HandleBackNavigation()` if compiler warnings appear.

---

### Positive Aspects ✅

- **Correct API detection**: `OperatingSystem.IsAndroidVersionAtLeast(33)`
- **Proper lifecycle management**: Register in `OnCreate`, unregister in `OnDestroy`
- **Excellent documentation**: Clear comments with Android documentation links
- **Code consolidation**: Unified handling reduces duplication
- **Backward compatibility**: Maintains support for Android 12 and earlier

---

### Testing Requirements

Given this is a late-stage fix with potentially limited testing, thorough validation is essential:

#### Platform Coverage
- ✅ **Android 12 and earlier**: Legacy `OnBackPressed()` path
- ✅ **Android 13+**: New predictive back gesture with animation

#### Lifecycle Scenarios
- ✅ Configuration changes (rotation, theme changes)
- ✅ Activity recreation (low memory, background/foreground)
- ✅ Memory leak verification (repeated lifecycle events)

#### Navigation Scenarios  
- ✅ Simple navigation (NavigationPage push/pop)
- ✅ Shell navigation
- ✅ Modal navigation
- ✅ Fragment back stack management
- ✅ Custom back handling via lifecycle events (`preventBackPropagation`)

#### Edge Cases
- ✅ Rapid back gestures
- ✅ Interrupted/cancelled back gestures
- ✅ Back gesture while navigation in progress

---

### Recommendation

**REQUEST CHANGES** - Please address the two issues above (double registration and missing dispose) before merging. These are important for stability and proper resource management.

The overall design is sound and correctly implements predictive back gesture support while maintaining MAUI's existing navigation architecture.

mattleibow reviewed (changes requested):
## Thorough Review: Predictive Back Gesture Support for Android 13+

### Summary
This PR adds predictive back gesture support for Android 13+ (API 33+) by implementing the `OnBackInvokedCallback` API. It centralizes back navigation handling in a new `HandleBackNavigation()` method used by both the legacy `OnBackPressed()` and the new predictive back callback system.

The approach correctly integrates MAUI's lifecycle event system with Android's predictive back gestures while maintaining backward compatibility.

---

### Critical Issues

#### 🔴 **BLOCKING: Potential Double Registration of Callback**
**Location**: `MauiAppCompatActivity.cs`, lines 44-45

The code could register the same callback multiple times if `OnCreate()` is called repeatedly during the activity lifecycle (e.g., configuration changes, activity recreation).

**Current Code**:
```csharp
_predictiveBackCallback ??= new PredictiveBackCallback(this);
dispatcher?.RegisterOnBackInvokedCallback(0, _predictiveBackCallback);
```

**Issue**: `??=` prevents creating a new instance, but `RegisterOnBackInvokedCallback()` is called unconditionally, potentially registering the same callback multiple times.

**Required Fix**:
```csharp
if (_predictiveBackCallback is null)
{
    _predictiveBackCallback = new PredictiveBackCallback(this);
    dispatcher?.RegisterOnBackInvokedCallback(0, _predictiveBackCallback);
}
```

---

#### ⚠️ **HIGH: Missing Dispose Call**
**Location**: `MauiAppCompatActivity.cs`, `OnDestroy()` around line 53

`PredictiveBackCallback` inherits from `Java.Lang.Object` and should be explicitly disposed to prevent memory leaks.

**Current Code**:
```csharp
OnBackInvokedDispatcher?.UnregisterOnBackInvokedCallback(_predictiveBackCallback);
_predictiveBackCallback = null;
```

**Required Fix**:
```csharp
OnBackInvokedDispatcher?.UnregisterOnBackInvokedCallback(_predictiveBackCallback);
_predictiveBackCallback.Dispose();
_predictiveBackCallback = null;
```

---

### Design Discussion

#### Regarding @mattleibow's Question: Should `base.OnBackPressed()` be called from PredictiveBackCallback?

**Answer: Yes, this is the correct approach.**

**Reasoning**:
1. **MAUI Architecture**: MAUI uses `AndroidLifecycle.OnBackPressed` events to allow application code to intercept and customize back navigation behavior
2. **Platform Behavior**: `base.OnBackPressed()` in `AppCompatActivity` handles critical platform operations:
   - Fragment back stack management
   - Activity finishing
   - Integration with AndroidX navigation components
3. **Compatibility**: This design maintains existing MAUI navigation patterns while adding predictive back animation support
4. **Control Flow**:
   - User performs back gesture → `OnBackInvoked()` called
   - `HandleBackNavigation()` invokes MAUI lifecycle events
   - App code can set `preventBackPropagation = true` to handle custom back behavior
   - If not prevented → `base.OnBackPressed()` performs default platform navigation

This unified approach correctly bridges MAUI's lifecycle system with Android's predictive back gesture API while preserving backward compatibility with Android 12 and earlier.

---

### Minor Issues

#### ℹ️ **LOW: Hardcoded Priority Value**
**Location**: `MauiAppCompatActivity.cs`, line 45

The priority `0` is hardcoded. While correct (`PRIORITY_DEFAULT = 0`), adding a comment would improve clarity:

```csharp
// Priority 0 = PRIORITY_DEFAULT: callback invoked only when no higher-priority callback handles the event
dispatcher?.RegisterOnBackInvokedCallback(0, _predictiveBackCallback);
```

#### 📝 **NOTE: Pragma Warnings**
The pragma warnings (`CA1416`/`CA1422`) were removed from `OnBackPressed()` when the code moved to `HandleBackNavigation()`. These warnings suppress obsolete API alerts. While the warnings aren't critical and were part of older Android code patterns, they could be re-added to `HandleBackNavigation()` if compiler warnings appear.

---

### Positive Aspects ✅

- **Correct API detection**: `OperatingSystem.IsAndroidVersionAtLeast(33)`
- **Proper lifecycle management**: Register in `OnCreate`, unregister in `OnDestroy`
- **Excellent documentation**: Clear comments with Android documentation links
- **Code consolidation**: Unified handling reduces duplication
- **Backward compatibility**: Maintains support for Android 12 and earlier

---

### Testing Requirements

Given this is a late-stage fix with potentially limited testing, thorough validation is essential:

#### Platform Coverage
- ✅ **Android 12 and earlier**: Legacy `OnBackPressed()` path
- ✅ **Android 13+**: New predictive back gesture with animation

#### Lifecycle Scenarios
- ✅ Configuration changes (rotation, theme changes)
- ✅ Activity recreation (low memory, background/foreground)
- ✅ Memory leak verification (repeated lifecycle events)

#### Navigation Scenarios  
- ✅ Simple navigation (NavigationPage push/pop)
- ✅ Shell navigation
- ✅ Modal navigation
- ✅ Fragment back stack management
- ✅ Custom back handling via lifecycle events (`preventBackPropagation`)

#### Edge Cases
- ✅ Rapid back gestures
- ✅ Interrupted/cancelled back gestures
- ✅ Back gesture while navigation in progress

---

### Recommendation

Please address the two issues above (double registration and missing dispose) before merging. These are important for stability and proper resource management.

The overall design is sound and correctly implements predictive back gesture support while maintaining MAUI's existing navigation architecture.

mattleibow committed:
Fix predictive back callback registration and resource management

- Fix potential double registration by checking if callback is null before creating/registering
- Add Dispose() call in OnDestroy() to prevent memory leaks
- Change field type to concrete PredictiveBackCallback for better performance (CA1859)
- Add comment explaining PRIORITY_DEFAULT value

These changes address code review feedback while maintaining the PR's core functionality.

mattleibow committed:
Simplify code to reduce nesting

mattleibow commented:
/azp run MAUI-UITests-public
